File: src/main/java/me/sd_master92/customvoting/subjects/VoteTopStand.java
package me.sd_master92.customvoting.subjects;

import me.sd_master92.customvoting.Main;
import me.sd_master92.customvoting.VoteFile;
import me.sd_master92.customvoting.constants.Data;
import me.sd_master92.customvoting.constants.Messages;
import me.sd_master92.customvoting.constants.Voter;
import me.sd_master92.customvoting.database.PlayerTable;
import me.sd_master92.customvoting.gui.GUI;
import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.Material;
import org.bukkit.World;
import org.bukkit.configuration.ConfigurationSection;
import org.bukkit.entity.ArmorStand;
import org.bukkit.entity.Entity;
import org.bukkit.entity.EntityType;
import org.bukkit.entity.Player;
import org.bukkit.inventory.EntityEquipment;
import org.bukkit.inventory.EquipmentSlot;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.SkullMeta;
import org.bukkit.scheduler.BukkitRunnable;

import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

public class VoteTopStand
{
    private static final Map<Integer, VoteTopStand> voteTops = new HashMap<>();

    private final Main plugin;
    private final int top;

    private ArmorStand topStand;
    private ArmorStand nameStand;
    private ArmorStand votesStand;

    public VoteTopStand(Main plugin, int top, Player player)
    {
        this.plugin = plugin;
        this.top = top;

        ConfigurationSection section = plugin.getData().getConfigurationSection(Data.VOTE_TOP_STANDS + "." + top);
        if (section != null)
        {
            if (player != null)
            {
                player.sendMessage(ChatColor.RED + "That Vote Stand already exists.");
                return;
            } else
            {
                topStand = getArmorStand(section.getString("top"));
                nameStand = getArmorStand(section.getString("name"));
                votesStand = getArmorStand(section.getString("votes"));
            }
        } else if (player != null)
        {
            create(player);
        }

        voteTops.put(top, this);
        update();
    }

    public VoteTopStand(Main plugin, int top)
    {
        this(plugin, top, null);
    }

    public static VoteTopStand get(int top)
    {
        return voteTops.get(top);
    }

    public static void updateAll(Main plugin)
    {
        if (voteTops.isEmpty())
        {
            initialize(plugin);
        }
        new BukkitRunnable()
        {
            @Override
            public void run()
            {
                for (VoteTopStand voteTop : voteTops.values())
                {
                    voteTop.update();
                }
            }
        }.runTaskLater(plugin, 40L);
    }

    private static void initialize(Main plugin)
    {
        ConfigurationSection section = plugin.getData().getConfigurationSection(Data.VOTE_TOP_STANDS);
        if (section != null)
        {
            for (String n : section.getKeys(false))
            {
                try
                {
                    int top = Integer.parseInt(n);
                    new VoteTopStand(plugin, top);
                } catch (Exception ignored)
                {

                }
            }
        }
    }

    private ArmorStand getArmorStand(String uuid)
    {
        if (uuid != null)
        {
            Entity entity =
                    plugin.getServer().getEntity(UUID.fromString(uuid));
            if (entity instanceof ArmorStand)
            {
                return (ArmorStand) entity;
            }
        }
        return null;
    }

    private void create(Player player)
    {
        World world = player.getLocation().getWorld();
        if (world != null)
        {
            topStand =
                    (ArmorStand) world.spawnEntity(player.getLocation().add(0, 1, 0),
                            EntityType.ARMOR_STAND);
            plugin.getData().set(Data.VOTE_TOP_STANDS + "." + top + ".top",
                    topStand.getUniqueId().toString());

            topStand.setVisible(false);
            topStand.setRemoveWhenFarAway(false);
            topStand.setSilent(true);
            topStand.setPersistent(true);
            topStand.setGravity(false);
            topStand.setCustomNameVisible(true);

            nameStand =
                    (ArmorStand) world.spawnEntity(player.getLocation().add(0, 0.5, 0),
                            EntityType.ARMOR_STAND);
            plugin.getData().set(Data.VOTE_TOP_STANDS + "." + top + ".name",
                    nameStand.getUniqueId().toString());

            nameStand.setVisible(false);
            nameStand.setRemoveWhenFarAway(false);
            nameStand.setSilent(true);
            nameStand.setPersistent(true);
            nameStand.setGravity(false);
            nameStand.setCustomNameVisible(true);

            votesStand = (ArmorStand) world.spawnEntity(player.getLocation(), EntityType.ARMOR_STAND);
            plugin.getData().set(Data.VOTE_TOP_STANDS + "." + top + ".votes",
                    votesStand.getUniqueId().toString());
            plugin.getData().saveConfig();

            votesStand.setRemoveWhenFarAway(false);
            votesStand.setSilent(true);
            votesStand.setPersistent(true);
            votesStand.setGravity(false);
            votesStand.setCustomNameVisible(true);

            votesStand.addEquipmentLock(EquipmentSlot.HEAD, ArmorStand.LockType.REMOVING_OR_CHANGING);
            votesStand.addEquipmentLock(EquipmentSlot.CHEST, ArmorStand.LockType.REMOVING_OR_CHANGING);
            votesStand.addEquipmentLock(EquipmentSlot.LEGS, ArmorStand.LockType.REMOVING_OR_CHANGING);
            votesStand.addEquipmentLock(EquipmentSlot.FEET, ArmorStand.LockType.REMOVING_OR_CHANGING);
            votesStand.addEquipmentLock(EquipmentSlot.HAND, ArmorStand.LockType.REMOVING_OR_CHANGING);
            EntityEquipment entityEquipment = votesStand.getEquipment();
            if (entityEquipment != null)
            {
                switch (top)
                {
                    case 1:
                        entityEquipment.setChestplate(GUI.createItem(Material.DIAMOND_CHESTPLATE, true));
                        entityEquipment.setLeggings(GUI.createItem(Material.DIAMOND_LEGGINGS, true));
                        entityEquipment.setBoots(GUI.createItem(Material.DIAMOND_BOOTS, true));
                        entityEquipment.setItemInMainHand(GUI.createItem(Material.DIAMOND_SWORD,
                                true));
                        break;
                    case 2:
                        entityEquipment.setChestplate(GUI.createItem(Material.GOLDEN_CHESTPLATE, true));
                        entityEquipment.setLeggings(GUI.createItem(Material.GOLDEN_LEGGINGS, true));
                        entityEquipment.setBoots(GUI.createItem(Material.GOLDEN_BOOTS, true));
                        entityEquipment.setItemInMainHand(GUI.createItem(Material.GOLDEN_SWORD,
                                true));
                        break;
                    case 3:
                        entityEquipment.setChestplate(GUI.createItem(Material.IRON_CHESTPLATE, true));
                        entityEquipment.setLeggings(GUI.createItem(Material.IRON_LEGGINGS, true));
                        entityEquipment.setBoots(GUI.createItem(Material.IRON_BOOTS, true));
                        entityEquipment.setItemInMainHand(GUI.createItem(Material.IRON_SWORD,
                                true));
                        break;
                    default:
                        entityEquipment.setChestplate(GUI.createItem(Material.CHAINMAIL_CHESTPLATE, true));
                        entityEquipment.setLeggings(GUI.createItem(Material.CHAINMAIL_LEGGINGS, true));
                        entityEquipment.setBoots(GUI.createItem(Material.CHAINMAIL_BOOTS, true));
                        entityEquipment.setItemInMainHand(GUI.createItem(Material.STONE_SWORD, true));
                }
            }
        }
        player.sendMessage(ChatColor.GREEN + "Registered Vote Stand #" + top);
    }

    private void update()
    {
        Voter voteFile = plugin.useDatabase() ? PlayerTable.getTopVoter(plugin, top) : VoteFile.getTopVoter(plugin,
                top);

        Map<String, String> placeholders = new HashMap<>();
        placeholders.put("%TOP%", "" + top);
        if (voteFile != null)
        {
            placeholders.put("%PLAYER%", voteFile.getName());
            placeholders.put("%VOTES%", "" + voteFile.getVotes());
        } else
        {
            placeholders.put("%PLAYER%", ChatColor.RED + "Unknown");
            placeholders.put("%VOTES%", "" + 0);
        }

        topStand.setCustomName(Messages.VOTE_TOP_STANDS_TOP.getMessage(plugin, placeholders));
        nameStand.setCustomName(Messages.VOTE_TOP_STANDS_CENTER.getMessage(plugin, placeholders));
        votesStand.setCustomName(Messages.VOTE_TOP_STANDS_BOTTOM.getMessage(plugin, placeholders));

        ItemStack skull = new ItemStack(Material.PLAYER_HEAD);
        SkullMeta skullMeta = (SkullMeta) skull.getItemMeta();
        if (skullMeta != null && voteFile != null)
        {
            try
            {
                skullMeta.setOwningPlayer(Bukkit.getOfflinePlayer(UUID.fromString(voteFile.getUuid())));
                skull.setItemMeta(skullMeta);
            } catch (Exception ignored)
            {
            }
        }
        EntityEquipment entityEquipment = votesStand.getEquipment();
        if (entityEquipment != null)
        {
            entityEquipment.setHelmet(skull);
        }
    }

    public void delete(Player player)
    {
        topStand.remove();
        nameStand.remove();
        votesStand.remove();

        plugin.getData().set(Data.VOTE_TOP_STANDS + "." + top, null);
        plugin.getData().saveConfig();

        voteTops.remove(top);

        player.sendMessage(ChatColor.GREEN + "Successfully deleted Vote Stand #" + top);
    }
}

File: src/main/kotlin/me/sd_master92/customvoting/subjects/VoteTopStand.kt
package me.sd_master92.customvoting.subjects

import me.sd_master92.core.tasks.TaskTimer
import me.sd_master92.customvoting.CV
import me.sd_master92.customvoting.constants.Voter
import me.sd_master92.customvoting.constants.enumerations.Data
import me.sd_master92.customvoting.constants.enumerations.Message
import me.sd_master92.customvoting.constants.enumerations.PMessage
import me.sd_master92.customvoting.gui.items.SimpleItem
import net.citizensnpcs.api.CitizensAPI
import net.citizensnpcs.api.npc.NPC
import org.bukkit.Bukkit
import org.bukkit.Location
import org.bukkit.Material
import org.bukkit.entity.ArmorStand
import org.bukkit.entity.EntityType
import org.bukkit.entity.LivingEntity
import org.bukkit.entity.Player
import org.bukkit.inventory.EntityEquipment
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.meta.SkullMeta
import java.util.*


class VoteTopStand constructor(private val plugin: CV, private val top: Int, player: Player? = null)
{
    private var topStand: ArmorStand? = null
    private var nameStand: ArmorStand? = null
    private var votesStand: ArmorStand? = null
    private var citizen: NPC? = null
    private val path = Data.VOTE_TOP_STANDS.path + ".$top"

    private fun registerArmorStands()
    {
        val section = plugin.data.getConfigurationSection(path)
        if (section != null)
        {
            topStand = getArmorStand(section.getString("top"))
            nameStand = getArmorStand(section.getString("name"))
            votesStand = getArmorStand(section.getString("votes"))
            if (!doesNotExist())
            {
                if (CV.CITIZENS)
                {
                    if (votesStand?.isVisible == true)
                    {
                        votesStand?.isVisible = false
                        votesStand?.equipment?.clear()
                    }
                    if (citizen == null)
                    {
                        citizen = if (section.getString("citizen") == null)
                        {
                            createCitizen(votesStand!!.location)
                        } else
                        {
                            CitizensAPI.getNPCRegistry().getByUniqueId(UUID.fromString(section.getString("citizen")))
                        }
                        if (citizen != null && !citizen!!.isSpawned)
                        {
                            citizen!!.spawn(plugin.data.getLocation(path))
                        }
                    }
                } else
                {
                    votesStand?.isVisible = true
                }
            } else
            {
                delete()
            }
        }
    }

    private fun getArmorStand(uuid: String?): ArmorStand?
    {
        if (uuid != null)
        {
            val entity = Bukkit.getEntity(UUID.fromString(uuid))
            if (entity is ArmorStand)
            {
                return entity
            }
        }
        return null
    }

    private fun create(player: Player)
    {
        val world = player.location.world
        if (world != null)
        {
            val topStand = spawnArmorStand(player.location.add(0.0, 1.0, 0.0))
            plugin.data["$path.top"] = topStand.uniqueId.toString()
            this.topStand = topStand

            val nameStand = spawnArmorStand(player.location.add(0.0, 0.5, 0.0))
            plugin.data["$path.name"] = nameStand.uniqueId.toString()
            this.nameStand = nameStand

            val votesStand = spawnArmorStand(player.location)
            plugin.data["$path.votes"] = votesStand.uniqueId.toString()
            plugin.data.saveConfig()
            this.votesStand = votesStand

            if (CV.CITIZENS)
            {
                citizen = createCitizen(player.location)
            } else
            {
                votesStand.isVisible = true
                val entityEquipment = votesStand.equipment
                if (entityEquipment != null)
                {
                    setEntityEquipment(entityEquipment)
                }
            }
        }
        plugin.data.setLocation(path, player.location)
        player.sendMessage(PMessage.VOTE_TOP_MESSAGE_STAND_CREATED_X.with("$top"))
    }

    private fun createCitizen(loc: Location, name: String? = null): NPC
    {
        val citizen =
            CitizensAPI.getNPCRegistry().createNPC(EntityType.PLAYER, name ?: PMessage.PLAYER_NAME_UNKNOWN.toString())
        plugin.data["$path.citizen"] = citizen.uniqueId.toString()
        citizen.isProtected = true
        citizen.data().setPersistent(NPC.Metadata.NAMEPLATE_VISIBLE, false)
        citizen.spawn(loc)
        if (citizen.isSpawned)
        {
            val entityEquipment = (citizen.entity as LivingEntity).equipment
            if (entityEquipment != null)
            {
                setEntityEquipment(entityEquipment)
            }
        }
        return citizen
    }

    private fun setEntityEquipment(equipment: EntityEquipment)
    {
        when (top)
        {
            1    ->
            {
                equipment.chestplate = SimpleItem(Material.DIAMOND_CHESTPLATE, true)
                equipment.leggings = SimpleItem(Material.DIAMOND_LEGGINGS, true)
                equipment.boots = SimpleItem(Material.DIAMOND_BOOTS, true)
                equipment.setItemInMainHand(SimpleItem(Material.DIAMOND_SWORD, true))
            }

            2    ->
            {
                equipment.chestplate = SimpleItem(Material.GOLDEN_CHESTPLATE, true)
                equipment.leggings = SimpleItem(Material.GOLDEN_LEGGINGS, true)
                equipment.boots = SimpleItem(Material.GOLDEN_BOOTS, true)
                equipment.setItemInMainHand(SimpleItem(Material.GOLDEN_SWORD, true))
            }

            3    ->
            {
                equipment.chestplate = SimpleItem(Material.IRON_CHESTPLATE, true)
                equipment.leggings = SimpleItem(Material.IRON_LEGGINGS, true)
                equipment.boots = SimpleItem(Material.IRON_BOOTS, true)
                equipment.setItemInMainHand(SimpleItem(Material.IRON_SWORD, true))
            }

            else ->
            {
                equipment.chestplate = SimpleItem(Material.CHAINMAIL_CHESTPLATE, true)
                equipment.leggings = SimpleItem(Material.CHAINMAIL_LEGGINGS, true)
                equipment.boots = SimpleItem(Material.CHAINMAIL_BOOTS, true)
                equipment.setItemInMainHand(SimpleItem(Material.STONE_SWORD, true))
            }
        }
    }

    private fun update()
    {
        val topVoter = Voter.getTopVoter(plugin, top)
        val placeholders: MutableMap<String, String> = HashMap()
        placeholders["%TOP%"] = "" + top
        if (topVoter != null)
        {
            placeholders["%PLAYER%"] = topVoter.name
            placeholders["%VOTES%"] = "${topVoter.votes}"
            placeholders["%MONTHLY_VOTES%"] = "${topVoter.monthlyVotes}"
            if (CV.CITIZENS && citizen != null && citizen!!.name != topVoter.name)
            {
                citizen!!.name = topVoter.name
                citizen!!.despawn()
                citizen!!.data()?.setPersistent(NPC.Metadata.NAMEPLATE_VISIBLE, false)
                citizen!!.spawn(plugin.data.getLocation(path))
            }
        } else
        {
            placeholders["%PLAYER%"] = PMessage.PLAYER_NAME_UNKNOWN_COLORED.toString()
            placeholders["%VOTES%"] = "0"
            placeholders["%MONTHLY_VOTES%"] = "0"
            if (CV.CITIZENS && citizen != null && citizen!!.name != PMessage.PLAYER_NAME_UNKNOWN.toString())
            {
                citizen!!.name = PMessage.PLAYER_NAME_UNKNOWN.toString()
                citizen!!.despawn()
                citizen!!.data()?.setPersistent(NPC.Metadata.NAMEPLATE_VISIBLE, false)
                citizen!!.spawn(plugin.data.getLocation(path))
            }
        }
        if (doesNotExist() || citizenDoesNotExist())
        {
            registerArmorStands()
        }
        if (!CV.CITIZENS)
        {
            val skull = ItemStack(Material.PLAYER_HEAD)
            val skullMeta = skull.itemMeta as SkullMeta?
            if (skullMeta != null && topVoter != null)
            {
                try
                {
                    val uuid = UUID.fromString(topVoter.uuid)
                    skullMeta.owningPlayer = Bukkit.getPlayer(uuid) ?: Bukkit.getOfflinePlayer(uuid)
                    skull.itemMeta = skullMeta
                } catch (ignored: Exception)
                {
                }
            }
            votesStand?.equipment?.helmet = skull
        }
        topStand?.customName = Message.VOTE_TOP_STANDS_TOP.getMessage(plugin, placeholders)
        nameStand?.customName = Message.VOTE_TOP_STANDS_CENTER.getMessage(plugin, placeholders)
        votesStand?.customName = Message.VOTE_TOP_STANDS_BOTTOM.getMessage(plugin, placeholders)
    }

    private fun doesNotExist(): Boolean
    {
        if (topStand == null
            || nameStand == null
            || votesStand == null
        )
        {
            return true
        }
        return false
    }

    private fun citizenDoesNotExist(): Boolean
    {
        if (CV.CITIZENS && (citizen == null || !citizen!!.isSpawned))
        {
            return true
        }
        return false
    }

    fun delete(player: Player? = null)
    {
        topStand?.remove()
        topStand = null
        nameStand?.remove()
        nameStand = null
        votesStand?.remove()
        votesStand = null
        citizen?.let { CitizensAPI.getNPCRegistry().deregister(it) }
        plugin.data.deleteLocation(path)
        plugin.data[path] = null
        plugin.data.saveConfig()
        voteTops.remove(top)
        player?.sendMessage(PMessage.VOTE_TOP_MESSAGE_STAND_DELETED_X.with("$top"))
    }

    companion object
    {
        private val voteTops: MutableMap<Int, VoteTopStand> = HashMap()

        operator fun get(top: Int): VoteTopStand?
        {
            return voteTops[top]
        }

        fun spawnArmorStand(loc: Location): ArmorStand
        {
            val stand = loc.world!!.spawnEntity(loc, EntityType.ARMOR_STAND) as ArmorStand
            stand.removeWhenFarAway = false
            stand.isSilent = true
            stand.setGravity(false)
            stand.isCustomNameVisible = true
            stand.isInvulnerable = true
            stand.isVisible = false
            return stand
        }

        fun updateAll(plugin: CV)
        {
            if (voteTops.isEmpty())
            {
                initialize(plugin)
            }
            TaskTimer.delay(plugin, 40)
            {
                for (voteTop in voteTops.values)
                {
                    voteTop.update()
                }
            }.run()
        }

        private fun initialize(plugin: CV)
        {
            val section = plugin.data.getConfigurationSection(Data.VOTE_TOP_STANDS.path)
            if (section != null)
            {
                for (n in section.getKeys(false))
                {
                    try
                    {
                        val top = n.toInt()
                        VoteTopStand(plugin, top)
                    } catch (ignored: Exception)
                    {
                    }
                }
            }
        }
    }

    init
    {
        val section = plugin.data.getConfigurationSection(path)
        if (section != null)
        {
            if (player != null)
            {
                player.sendMessage(PMessage.GENERAL_ERROR_ALREADY_EXIST_X.with(PMessage.VOTE_TOP_UNIT_STAND.toString()))
            } else
            {
                registerArmorStands()
                voteTops[top] = this
                update()
            }
        } else player?.let {
            create(it)
            voteTops[top] = this
            update()
        }
    }
}